Pr implement reconcile certs

sbueringer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

hidekazuna committed:
Add MachineDeployment template (#384)

This patch adds MachineDeployment template as in
cluster-api-provider-aws.

prankul88 committed:
Improvements in config.md file (#379)

sbueringer committed:
Reconcile certificates and store them in the cluster crd

This is heavily inspired bei CAPA. We now reconcile and store certificates
in the Cluster CRD. Thus it's possible to distribute the same CAs over all
control plane nodes (as soon as multi-node control plane supported is
implemented). We also don't have to ssh on the/a control plane node to get
a valid kubeconfig. We now can just generate one from the CA.

sbueringer committed:
Generate a kubeconfig for the Workload cluster instead of reusing the controllerClient

This enables us to run the controller outside the Workload Cluster, e.g. in a Management
Cluster.

sbueringer committed:
Add options for local user data

Now it's possible to use a local userdata folder. This is mostly useful
for development to avoid updating the user data Secrets all the time.

jichenjc committed:
move from deployment to statefulset (#376)

per discussion , deployment has sideeffect such as potential
race condition, so need switch to statefulset.

jichenjc committed:
Fix centos create generate issue (#390)

centos generate has some issue, missed  commands=>files changes

jichenjc commented on pkg/cloud/openstack/cluster/actuator.go:
```diff
@@ -66,12 +67,20 @@ func (a *Actuator) Reconcile(cluster *clusterv1.Cluster) error {
+	certificatesService := certificates.NewService()
+
 	// Load provider spec & status.
 	clusterProviderSpec, clusterProviderStatus, err := providerv1.ClusterSpecAndStatusFromProviderSpec(cluster)
```
ClusterSpecAndStatusFromProviderSpec    is this implemented?

sbueringer replied:
Yup here: https://github.com/sbueringer/cluster-api-provider-openstack/blob/pr-implement-reconcile-certs/pkg/apis/openstackproviderconfig/v1alpha1/register.go#L41

If it's not in the diff, it's because I implemented it in the PR before.

jichenjc replied:
ok, got it, thanks ~

jichenjc commented on pkg/apis/openstackproviderconfig/v1alpha1/types.go:
```diff
@@ -214,6 +211,30 @@ type OpenstackClusterProviderSpec struct {
+type KeyPair struct {
+	// base64 encoded cert and key
+	Cert []byte `json:"cert"`
+	Key  []byte `json:"key"`
```
consider to add omitempty here?

sbueringer replied:
Done

jichenjc commented on pkg/cloud/openstack/services/certificates/certificates.go:
```diff
@@ -0,0 +1,334 @@
+
+// ReconcileCertificates generate certificates if none exists.
+func (s *Service) ReconcileCertificates(clusterName string, clusterProviderSpec *v1alpha1.OpenstackClusterProviderSpec) error {
+	if !clusterProviderSpec.CAKeyPair.HasCertAndKey() {
```
from code, looks like you are trying to generate those keys? if it's generated, then how do you store them and reuse? looks to me the Spec are the desired state but if you didn't give key at beginning, it's not desired state?? just curious ..

sbueringer replied:
I'll move this PR and your comment over to the kubernetes-sigs repo now that the first PR is merged :)

sbueringer replied:
Moved here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/394/files#r302672418

jichenjc committed:
Fix klog issue (#388)

upgrade klog from 0.1 to 0.3 to avoid no log issue

sbueringer committed:
factored out services (#383)

* factored out services

These changes are heavily inspired by CAPA. They should improve the structure
of the code for the upcoming addition of multi-node control plane
support.

* fix tests

Commit:
Added Stefan Büringer (sbueringer) to be a reviewer (#393)

Stefan works currently on cleaning up stuff as well as the multi master
implementation. I want him to be able to review PRs.

sbueringer commented:
Moved to https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/394
